connectors/src/connectors/google_drive/temporal/spreadsheets.ts

```diff
@@ -0,0 +1,328 @@
+import type { ModelId } from "@dust-tt/types";
+import { slugify } from "@dust-tt/types";
+import { stringify } from "csv-stringify/sync";
+import type { sheets_v4 } from "googleapis";
+import { google } from "googleapis";
+import type { OAuth2Client } from "googleapis-common";
+
+import { dataSourceConfigFromConnector } from "@connectors/lib/api/data_source_config";
+import { concurrentExecutor } from "@connectors/lib/async_utils";
+import { deleteTable, upsertTableFromCsv } from "@connectors/lib/data_sources";
+import type { GoogleDriveFiles } from "@connectors/lib/models/google_drive";
+import { GoogleDriveSheet } from "@connectors/lib/models/google_drive";
+import { connectorHasAutoPreIngestAllDatabasesFF } from "@connectors/lib/workspace";
+import logger from "@connectors/logger/logger";
+import { ConnectorResource } from "@connectors/resources/connector_resource";
+import type { GoogleDriveObjectType } from "@connectors/types/google_drive";
+
+const MAXIMUM_NUMBER_OF_GSHEET_ROWS = 1000;
+
+type Sheet = sheets_v4.Schema$ValueRange & {
+  id: number;
+  spreadsheet: {
+    id: string;
+    title: string;
+  };
+  title: string;
+};
+
+export function isGoogleDriveSpreadSheetFile(
+  file: GoogleDriveObjectType | GoogleDriveFiles
+) {
+  return file.mimeType === "application/vnd.google-apps.spreadsheet";
+}
+
+function makeTableIdFromSheetId(spreadsheetId: string, sheetId: number) {
+  return `google-spreadsheet-${spreadsheetId}-sheet-${sheetId}`;
+}
+
+async function upsertSheetInDb(connector: ConnectorResource, sheet: Sheet) {
+  await GoogleDriveSheet.upsert({
+    connectorId: connector.id,
+    driveFileId: sheet.spreadsheet.id,
+    driveSheetId: sheet.id,
+    name: sheet.title,
+  });
+}
+
+async function upsertTable(
+  connector: ConnectorResource,
+  sheet: Sheet,
+  rows: string[][]
+) {
+  const dataSourceConfig = await dataSourceConfigFromConnector(connector);
+
+  const { id, spreadsheet, title } = sheet;
+  // Table name will be slugify in front.
+  const tableName = `${spreadsheet.title} - ${title}`;
+  const tableDescription = `Structured data from the Google Spreadsheet (${spreadsheet.title}) and sheet (${title}`;
+
+  const csv = stringify(rows);
+
+  // Upserting is safe: Core truncates any previous table with the same Id before
+  // the operation. Note: Renaming a sheet in Google Drive retains its original Id.
+  await upsertTableFromCsv({
+    dataSourceConfig,
+    tableId: makeTableIdFromSheetId(spreadsheet.id, id),
+    tableName,
+    tableDescription,
+    tableCsv: csv,
+    loggerArgs: {
+      connectorId: connector.id,
+      sheetId: id,
+      spreadsheetId: spreadsheet.id,
+    },
+  });
+}
+
+function getValidRows(allRows: string[][], loggerArgs: object) {
```

Comment:
Yeah I need to play a bit more with the GSheet API since it seems very powerful and can probably help us on this aspect. I'll tackle this in a follow up PR.

Reply:
Sounds good !

Reply:
So for now if the first lines are empty, we will simply ignore. It's not perfect and will get polished in the next few iterations.

Reply:
Ok updated the logic in [f1be39b](https://github.com/dust-tt/dust/pull/3769/commits/f1be39bbf7c1426f20610ea258ce3d6337c604bf) to bring more resiliency.